run_test.py

```diff
@@ -0,0 +1,229 @@
+
+
+def prepare_environment(verbose):
+    print(f"{YELLOW} --- Prepare Environment --- {NC}")
+
+    test_dir = "test_por"
+    devtools_repo = "ssh://git.phoesys.com:29418/devtools"
+    devtools_dir = "devtools"
+
+    if os.path.isdir(test_dir):
+        print(f"Directory '{test_dir}' exists. Removing...")
+        subprocess.run(["rm", "-rf", test_dir], check=True)
+
+    if os.path.isdir(devtools_dir):
+        print(f"Directory '{devtools_dir}' exists. Removing...")
+        subprocess.run(["rm", "-rf", devtools_dir], check=True)
+
+    print(f"Cloning '{devtools_repo}' into '{devtools_dir}'...")
+    clone_cmd = ["git", "clone", devtools_repo, devtools_dir]
+    if verbose:
+        subprocess.run(clone_cmd, check=True)
+    else:
+        subprocess.run(clone_cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=True)
+
+    venv_dir = "_venv"
+    if os.path.isdir(venv_dir):
+        print(f"{YELLOW}Removing existing virtual environment '{venv_dir}'{NC}")
+        subprocess.run(["rm", "-rf", venv_dir], check=True)
+
+    print(f"{YELLOW}Creating new virtual environment '{venv_dir}'{NC}")
+
+    subprocess.run(["./utils/prepare-venv.py"], check=True)
+
+    for root, dirs, files in os.walk("."):
+        if "requirements.txt" in files:
+            requirements_path = os.path.join(root, "requirements.txt")
+            print(f"Installing dependencies from '{requirements_path}'")
+            try:
+                subprocess.run([os.path.join(venv_dir, "bin", "pip"), "install", "-r", requirements_path], check=True)
+            except subprocess.CalledProcessError as e:
+                print(f"Failed to install dependencies from '{requirements_path}': {e}")
+
+    print(f"{YELLOW}Environment preparation completed{NC}")
+
+
+def run_pytest_tests(search_dirs_pytest, verbose):
+    print(f"{YELLOW} --- Searching for pytest tests --- {NC}")
+    exit_code = 0
+
+    for folder in search_dirs_pytest:
+        pytest_cmd = ["./_venv/bin/python3", "-m", "pytest"]
+
+        if not verbose:
+            pytest_cmd.append("--tb=no")
+
+        pytest_cmd.append(folder)
+
+        print(f"Running pytest in directory: {folder}")
+
+        try:
+            subprocess.run(pytest_cmd, check=True)
+        except subprocess.CalledProcessError as e:
+            print("Pytest failed with error:")
+            print(e.output)
+            exit_code = e.returncode
+
+    print(f"{YELLOW} --- Pytest tests completed --- {NC}")
+    return exit_code
+
+
+def run_binary_tests(search_dirs_bin, target, verbose):
+    print(f"{YELLOW} --- Start binary tests --- {NC}")
+
+    build_dir = f"_build/{target}/prog"
+    found_binaries = []
+    exit_code = 0
+
+    for search_dir in search_dirs_bin:
+        for root, dirs, files in os.walk(search_dir):
+            for file in files:
+                if file == "Makefile":
+                    with open(os.path.join(root, file), "r") as f:
+                        for line in f:
```

Comment:
Guard clause would save you one layer of indentation here
```
if file != "Makefile":
    continue
with open(...
```
Also I'd try exporting three FOR loops to another function
```
def all_files():
    res = []
    forforfor:
                                 res.append(file)
    return res
...

for file in all_files():
...
```